daemons/based/pacemaker-based.c

```diff
@@ -52,141 +55,149 @@ static bool startCib(const char *filename);
 extern int write_cib_contents(gpointer p);
 void cib_cleanup(void);
 
+static crm_exit_t exit_code = CRM_EX_OK;
+
 static void
 cib_enable_writes(int nsig)
 {
     crm_info("(Re)enabling disk writes");
     cib_writes_enabled = TRUE;
 }
 
-static pcmk__cli_option_t long_options[] = {
-    // long option, argument type, storage, short option, description, flags
-    {
-        "help", no_argument, 0, '?',
-        "\tThis text", pcmk__option_default
-    },
-    {
-        "verbose", no_argument, NULL, 'V',
-        "\tIncrease debug output", pcmk__option_default
-    },
-    {
-        "stand-alone", no_argument, NULL, 's',
-        "\tAdvanced use only", pcmk__option_default
-    },
-    {
-        "disk-writes", no_argument, NULL, 'w',
-        "\tAdvanced use only", pcmk__option_default
-    },
-    {
-        "cib-root", required_argument, NULL, 'r',
-        "\tAdvanced use only", pcmk__option_default
-    },
-    { 0, 0, 0, 0 }
+static gboolean
+stand_alone_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+               GError **error)
+{
+    int rc = 0;
+    struct passwd *pwentry = NULL;
+
+    stand_alone = TRUE;
+    preserve_status = TRUE;
+    cib_writes_enabled = FALSE;
+
+    pwentry = getpwnam(CRM_DAEMON_USER);
+    CRM_CHECK(pwentry != NULL,
+              exit_code = CRM_EX_FATAL;
+              g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                          "Invalid uid (%s) specified", CRM_DAEMON_USER);
+              return FALSE);
+
+    rc = setgid(pwentry->pw_gid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not set group to %d", pwentry->pw_gid);
+        return FALSE;
+    }
+
+    rc = initgroups(CRM_DAEMON_USER, pwentry->pw_gid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not setup groups for user %d", pwentry->pw_uid);
+        return FALSE;
+    }
+
+    rc = setuid(pwentry->pw_uid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not set user to %d", pwentry->pw_uid);
+        return FALSE;
+    }
+    return TRUE;
+}
+
+static GOptionEntry entries[] = {
+    { "stand-alone", 's', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
+      stand_alone_cb, "(Advanced use only) Run in stand-alone mode", NULL },
+
+    { "disk-writes", 'w', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
+      &cib_writes_enabled,
+      "(Advanced use only) Enable disk writes (enabled by default unless in "
+      "stand-alone mode)", NULL },
+
+    { "cib-root", 'r', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &cib_root,
```

Comment:
It seems like the "Using custom config location" notice message and the writability check might be better suited for a GOption callback, instead of just assigning here and doing the check in `main()`. However, that does mean we would exit before destroying an old instance, if that matters.